netmiko/cisco/cisco_vxr_ssh.py

```diff
@@ -0,0 +1,634 @@
+from netmiko.cafy_custom_exceptions import SessionDownException, PromptNotFoundException, PatternNotFoundException
+from netmiko.cafy_custom_exceptions import ConfigCommitError, ConfigModeEnterError, ConfigModeExitError, LOOP_DELAY
+
+work_dir = os.getenv('CAFYKIT_WORK_DIR')
+if work_dir:
+    formatter = logging.Formatter(fmt='%(asctime)s %(levelname)-8s %(message)s',
+                                  datefmt='%Y-%m-%d %H:%M:%S')
+    paramiko_log_folder = os.path.join(work_dir, 'test_paramiko.log')
+    paramiko.util.log_to_file(paramiko_log_folder, level="DEBUG")
+    fh = logging.FileHandler(paramiko_log_folder)
+    fh.setFormatter(formatter)
+    log = logging.getLogger("netmiko")
+    log.addHandler(fh)
+    log.setLevel(logging.DEBUG)
+    log.addHandler(fh)
+else:
+    from netmiko import log
+
+DELAY_FACTOR_DEPR_SIMPLE_MSG = """\n
+Delay Factor is not used in Cisco VXR SSH Netmiko Library.
+Netmiko 4.x and later has deprecated the use of delay_factor.
+You should remove any use of delay_factor=x from this method call.\n"""
+
+MAX_LOOPS_DEPR_SIMPLE_MSG = """\n
+Max Loops is not used in Cisco VXR SSH Netmiko Library.
+Netmiko 4.x and later has deprecated the use of max_loops.
+You should remove any use of max_loops=x from this method call.\n"""
+
+
+class CiscoVxrSSH(CiscoXrSSH):
+    """
+    CiscoVxrSSH is based of CiscoXrSSH -- CiscoBaseConnection
+    """
+
+    def __init__(self, **kwargs):
+        """Constructor
+        """
+        # 30 minutes
+        self.max_read_timeout = kwargs.get('max_read_timeout', 1800)
+        # In netmiko4, max_read_timeout which was added in netmiko2's vxr_ssh doesnt exist,
+        # instead read_timeout serves the same purpose, therefore assigning max_read_timeout to read_timeout
+        self.read_timeout = self.max_read_timeout
+        kwargs["blocking_timeout"] = self.max_read_timeout
+        super().__init__(**kwargs)
+
+    def write_channel(self, out_data):
+        """Generic handler that will write to both SSH and telnet channel.
+
+        :param out_data: data to be written to the channel
+        :type out_data: str (can be either unicode/byte string)
+        """
+        # Netmiko4 already has @lock_channel decorator on write_channel() in base_Connection.py
+        # Therefore below self._lock_netmiko_session() is commented
+        #self._lock_netmiko_session()
+        try:
+            # Netmiko4 base_connection doesnt have _write_channel()
+            # Therefore calling super().write_channel()  which has same functionality as netmiko2's _write_channel
+            # self._write_channel(out_data)
+            super().write_channel(out_data)
+        except socket.error:
+            msg = "Session went down while writing command on channel. Command: {}".format(out_data)
+            log.error(msg)
+            raise SessionDownException(msg)
+        finally:
+            # Always unlock the SSH channel, even on exception.
+            self._unlock_netmiko_session()
```

Comment:
is this needed since write_channel() itself has a decorator doing lock, unlock

Reply:
good catch, i have removed this now.